src/libmongoc/src/mongoc/mongoc-client.c

```diff
@@ -1288,6 +1289,16 @@ mongoc_client_start_session (mongoc_client_t *client,
       csid = (uint32_t) _mongoc_rand_simple (&client->csid_rand_seed);
    } while (mongoc_set_get (client->client_sessions, csid));
 
+   /* causal consistency and snapshot cannot both be set. */
+   if (opts && mongoc_session_opts_get_causal_consistency (opts) &&
+       mongoc_session_opts_get_snapshot (opts)) {
+      bson_set_error (
+         error,
+         MONGOC_ERROR_CLIENT,
+         MONGOC_ERROR_CLIENT_SESSION_FAILURE,
+         "Only one of causal consistency and snapshot can be enabled.");
+      RETURN (NULL);
```

Comment:
To prevent leaking `ss`:
```suggestion
      _mongoc_client_push_server_session (client, ss);
      RETURN (NULL);
```

Reply:
Ah thanks, that should fix some of the test failures. Added.